code-dot-org/code-dot-org
# Gen AI lab: save student customizations to S3, output changes to chat window

Two major changes:
- stores (and loads) student model customizations from S3. If no student-specific customizations exist, we use levelbuilder customizations.
- on save, compares to the most recently saved version, and outputs the changed fields to the chat window.

**Example message logged on save (UI obviously needs work :))**

![image](https://github.com/code-dot-org/code-dot-org/assets/25372625/f216a968-9c04-45c6-86ac-8c84d6d0996e)

## Links

- Log to chat window: https://codedotorg.atlassian.net/browse/LABS-517
- Project saving: https://codedotorg.atlassian.net/browse/LABS-537

## Testing story

When testing manually, I was able to:
- load initial model configuration from levelbuilder.
- update that configuration, and hit the "Update"button. I see writes to S3 in the network tab.
- see a message written to the chat window with the appropriate fields marked as having been updated.
- reload the page, and see the newly saved model configuration loaded correctly.

## Follow-up work

- Styling improvements on messages sent to chat window.
